<div style="border:solid green 2px; padding: 20px"> <h1 style="color:green; margin-bottom:20px">Reviewers comment v1</h1>

Hello Elvis!

Im happy to review your project today 🙌

You can find my comments under the heading **«Review»**. I will categorize my comments in green, blue or red boxes like this:

<div class="alert alert-success">
    <b>Success:</b> if everything is done successfully
</div>
<div class="alert alert-warning">
    <b>Remarks:</b> if I can give some recommendations or ways to improve the project
</div>
<div class="alert alert-danger">
    <b>Needs fixing:</b> if the block requires some corrections. Work cant be accepted with the red comments
</div>

Please dont remove my comments :) If you have any questions dont hesitate to respond to my comments in a different section. 
<div class="alert alert-info"> <b>Student comments:</b> For example like this</div>    



 <div style="border:solid green 2px; padding: 20px">
<b>Reviewer's comment v1:</b>
    
<b>Overall Feedback</b> 
    

Hello Elvis,

Another project successfully completed - well done! 🏆 Your consistent effort and progress are truly commendable.

Our team is here to help you keep pushing forward and honing your skills as you advance through the program.

You’ll find general comments below in the notebook in the `Reviewer's comment v1:` blocks.

**What Was Great:**

- Data Loading and Preparation: You successfully loaded the dataset and correctly separated the features and target variable. This is a crucial first step, and you handled it appropriately.
- Model Training and Evaluation: Using a loop to train and evaluate each model is an efficient approach. You correctly used the accuracy_score metric to assess the models and identified the Random Forest classifier as the best-performing model.
- Test Set Evaluation: You validated the performance of your selected model on a test set, ensuring that the model generalizes well to unseen data. The test accuracy being close to the validation accuracy is a good sign.
   
    
Congratulations again on your accomplishment! Each project you complete adds to your growing expertise, and it’s exciting to see you make such great strides. Keep up the great work! 🎯

# TRIPLETEN PROJECT SPRINT  7
## Project Objectives

* Develop a model for Megaline to classify subscribers.
* Analyze behavior data for plan recommendations.
* Maximize model accuracy for plan prediction.

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
Great that you've added a short information about the project goal. It gives an overview of what you are going to achieve in this project.


In [1]:
# importing required libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [2]:
# Loading  the data
df = pd.read_csv('/datasets/users_behavior.csv')
df.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

<div class="alert alert-block alert-danger">
<b>Reviewer's comment v0</b>
    
Here you have a path to your local dataset which will not work in the Tripleten environment. Correct path to the file is the following: `/datasets/users_behavior.csv`

You can adjust your try/except structure with different paths so it will work everywhere.
    
Also, I recommend that you always run your project once more on the TripleTen platform before submitting your work to avoid possible errors. You can do this by selecting `Kernel -> Restart & Run All`.

<div class="alert alert-info"> <b>Student comments V0:</b> Thanks, I have used the required path,restarted and ran ALL</div>

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
👏

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


**From the above data info, there are no missing values**

<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
Well done! Data have been successfully loaded and inspected.

**2. Spliting the Source Data**

**I will split the data into training, validation, and test sets in the ratio 3:1:1**

In [4]:
# Separate features and target
features = df.drop('is_ultra', axis=1)
target = df['is_ultra']

# Split into training, validation, and test sets
features_train, features_test, target_train, target_test = train_test_split(
    features, target, test_size=0.2, random_state=12345
)

features_train, features_valid, target_train, target_valid = train_test_split(
    features_train, target_train, test_size=0.25, random_state=12345
)

# Check the sizes of the resulting datasets
print("Training set size:", len(features_train))
print("Validation set size:", len(features_valid))
print("Test set size:", len(features_test))

Training set size: 1928
Validation set size: 643
Test set size: 643


<div class="alert alert-warning">
<b>Reviewer's comment v1:</b>
    
Consider using stratified splits (`stratify` parameter) to ensure that class distributions are similar across training, validation, and test sets.



**3. Investigate the Quality of Different Models**


In [5]:
# Decision Tree
best_tree_accuracy = 0
best_tree_depth = 0
for depth in range(1, 11):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model.fit(features_train, target_train)
    predictions_valid = model.predict(features_valid)
    accuracy = accuracy_score(target_valid, predictions_valid)
    if accuracy > best_tree_accuracy:
        best_tree_accuracy = accuracy
        best_tree_depth = depth

print(f"Best Decision Tree Accuracy: {best_tree_accuracy}, Depth: {best_tree_depth}")

# Random Forest
best_forest_accuracy = 0
best_forest_est = 0
for est in range(10, 51, 10):
    model = RandomForestClassifier(random_state=12345, n_estimators=est)
    model.fit(features_train, target_train)
    predictions_valid = model.predict(features_valid)
    accuracy = accuracy_score(target_valid, predictions_valid)
    if accuracy > best_forest_accuracy:
        best_forest_accuracy = accuracy
        best_forest_est = est

print(f"Best Random Forest Accuracy: {best_forest_accuracy}, Estimators: {best_forest_est}")

# Logistic Regression
model = LogisticRegression(random_state=12345, solver='liblinear')
model.fit(features_train, target_train)
predictions_valid = model.predict(features_valid)
logistic_accuracy = accuracy_score(target_valid, predictions_valid)

print(f"Logistic Regression Accuracy: {logistic_accuracy}")

Best Decision Tree Accuracy: 0.7744945567651633, Depth: 7
Best Random Forest Accuracy: 0.7947122861586314, Estimators: 50
Logistic Regression Accuracy: 0.7293934681181959


<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
Everything is correct here! Great that you've managed to check multiple models. 

Some possible minor improvements: 

- If you find yourself repeating similar blocks of code, consider writing a function. This will make your code more organized and easier to maintain.
- Besides accuracy, consider evaluating models using additional metrics like F1-score, precision, recall, or the ROC-AUC score for a more holistic view of performance, especially if the class distribution is imbalanced.
- Consider using GridSearchCV or RandomizedSearchCV from sklearn.model_selection for a more systematic hyperparameter search.

**Findings:**

* Decision Tree: The accuracy varied with the max_depth. We found the best depth that maximized accuracy on the validation set.
* Random Forest: The accuracy also varied with the number of n_estimators. We found the optimal number of trees.
* Logistic Regression: Achieved a certain accuracy without hyperparameter tuning, but may not be the optimal model.

**4. Check the Quality of the Model Using the Test Set**

In [6]:
# Select the best model (based on validation accuracy)
if best_forest_accuracy > best_tree_accuracy and best_forest_accuracy > logistic_accuracy:
    best_model = RandomForestClassifier(random_state=12345, n_estimators=best_forest_est)
elif best_tree_accuracy > logistic_accuracy:
    best_model = DecisionTreeClassifier(random_state=12345, max_depth=best_tree_depth)
else:
    best_model = LogisticRegression(random_state=12345, solver='liblinear')

# Train the best model on the combined training and validation sets
features_train_full = pd.concat([features_train, features_valid])
target_train_full = pd.concat([target_train, target_valid])

best_model.fit(features_train_full, target_train_full)

# Make predictions on the test set
predictions_test = best_model.predict(features_test)

# Calculate the test accuracy
test_accuracy = accuracy_score(target_test, predictions_test)

print(f"Test Accuracy of the Best Model: {test_accuracy}")

Test Accuracy of the Best Model: 0.7807153965785381


<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
Great! Well above the required threshold. 

**5. Additional Task: Sanity Check the Model**

In [7]:
# Baseline: Predict the most frequent class
from sklearn.dummy import DummyClassifier

dummy_clf = DummyClassifier(strategy="most_frequent")
dummy_clf.fit(features_train_full, target_train_full)
dummy_predictions = dummy_clf.predict(features_test)
dummy_accuracy = accuracy_score(target_test, dummy_predictions)

print(f"Baseline Accuracy (Most Frequent Class): {dummy_accuracy}")

if test_accuracy > dummy_accuracy:
    print("Model performs better than the baseline.")
else:
    print("Model does not perform better than the baseline.")

Baseline Accuracy (Most Frequent Class): 0.6951788491446346
Model performs better than the baseline.


<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
Everything is correct here. 

# Conclusion
**A Random Forest model, with optimized hyperparameters, achieved 78.07% accuracy in predicting Megaline's mobile plan recommendations, exceeding the 75% threshold and outperforming a baseline, making it suitable for deployment.**


<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    

Great job on your overall conclusions and recommendations!  Your recommendations are well-thought and could be very valuable to the business.